core/trino-main/src/main/java/io/trino/execution/SqlTask.java

```diff
@@ -290,7 +292,16 @@ public Optional<Set<CatalogHandle>> getCatalogs()
     public boolean setCatalogs(Set<CatalogHandle> catalogs)
     {
         requireNonNull(catalogs, "catalogs is null");
-        return this.catalogs.compareAndSet(null, requireNonNull(catalogs, "catalogs is null"));
+        return this.catalogs.compareAndSet(null, catalogs);
+    }
+
+    public void setCatalogsLoaded(ListenableFuture<Void> catalogsLoadedFuture)
+    {
+        requireNonNull(catalogsLoadedFuture, "catalogsLoaded is null");
+        if (this.catalogsLoadedFuture.compareAndSet(null, catalogsLoadedFuture)) {
+            catalogsLoadedFuture.addListener(taskStateMachine::transitionToRunning, directExecutor());
```

Comment:
I think this will be triggered on failure on future too. You do not want that.

Replace those two lines with:
```java
            addCallback(catalogsLoadedFuture, new FutureCallback<>()
            {
                @Override
                public void onSuccess(Void result)
                {
                    taskStateMachine.transitionToRunning();
                }

                @Override
                public void onFailure(Throwable t)
                {
                    taskStateMachine.failed(t);
                }
            }, directExecutor());
```